config.org
* Package Management 
#+BEGIN_SRC emacs-lisp
(package-initialize)
(setq package-check-signature nil)
(when (>= emacs-major-version 24)
     (require 'package)
     (package-initialize)
     (setq package-archives '(("gnu"   . "http://elpa.emacs-china.org/gnu/")
                      ("melpa" . "http://elpa.emacs-china.org/melpa/"))))

 ;; cl - Common Lisp Extension
 (require 'cl)

 ;; Add Packages
 (defvar my/packages '(
                company
                evil
                nlinum-relative
                yasnippet
                ox-hugo
                helm
                monokai-theme
                hungry-delete
                swiper
                counsel
                popwin
                evil-collection
                slime
                paredit
                geiser
                ) "Default packages")

 (setq package-selected-packages my/packages)

 (defun my/packages-installed-p ()
     (loop for pkg in my/packages
           when (not (package-installed-p pkg)) do (return nil)
           finally (return t)))

 (unless (my/packages-installed-p)
     (message "%s" "Refreshing package database...")
     (package-refresh-contents)
     (dolist (pkg my/packages)
       (when (not (package-installed-p pkg))
         (package-install pkg))))
#+END_SRC 
* UI
** Tool Bar & Menu Bar & Scroll Bar
#+BEGIN_SRC emacs-lisp
(tool-bar-mode -1)
(menu-bar-mode -1)
(scroll-bar-mode -1)
#+END_SRC
** Welcome Screen
#+BEGIN_SRC emacs-lisp
(setq inhibit-splash-screen 1)
#+END_SRC
** Relative Line Number
#+BEGIN_SRC emacs-lisp
(require 'nlinum-relative)
(nlinum-relative-setup-evil)                    ;; setup for evil
(add-hook 'prog-mode-hook 'nlinum-relative-mode)
(add-hook 'text-mode-hook 'nlinum-relative-mode)
(setq nlinum-relative-current-symbol "")      ;; or "" for display current line number
(setq nlinum-relative-redisplay-delay 0)      ;; delay
(setq nlinum-relative-offset 0)                 ;; 1 if you want 0, 2, 3...
#+END_SRC
* Keybindings
** Evil
#+BEGIN_SRC emacs-lisp
(setq evil-want-keybinding nil)
(setq evil-want-C-i-jump nil) ;; use tab in normal mode while editing org file
(evil-mode 1) 
#+END_SRC
** Evil Collection
#+BEGIN_SRC emacs-lisp
(evil-collection-init)
#+END_SRC
** Ivy Mode
#+BEGIN_SRC emacs-lisp
(ivy-mode 1)
#+END_SRC
** Recentf
#+BEGIN_SRC emacs-lisp
(require 'recentf)
(recentf-mode 1)
(setq recentf-max-menu-items 25)
(global-set-key "\C-x\ \C-r" 'recentf-open-files)
#+END_SRC
** Shortcuts
*** F2 to Open Config
#+BEGIN_SRC emacs-lisp
(defun open-init-file()
  (interactive)
  (find-file "~/dotfiles/config.org"))
(global-set-key (kbd "<f2>") 'open-init-file)
#+END_SRC
*** Ctrl-w to delelte word
#+BEGIN_SRC emacs-lisp
(global-set-key (kbd "C-w") 'backward-kill-word)
#+END_SRC
*** Ctrl-h as Backspace
#+BEGIN_SRC emacs-lisp
(global-set-key (kbd "C-h") 'delete-backward-char)
#+END_SRC
*** Ctrl-s as Swiper
#+BEGIN_SRC emacs-lisp
(global-set-key "\C-s" 'swiper)
#+END_SRC
*** Meta-x as consel-M-x
#+BEGIN_SRC emacs-lisp
(global-set-key (kbd "M-x") 'counsel-M-x)
#+END_SRC
*** Ctrl-c p f as consel-git
#+BEGIN_SRC emacs-lisp
(global-set-key (kbd "C-c p f") 'counsel-git)
#+END_SRC
*** Ctrl-g as esc
**** function define
#+BEGIN_SRC emacs-lisp
(defun evil-keyboard-quit ()
  "Keyboard quit and force normal state."
  (interactive)
  (and evil-mode (evil-force-normal-state))
  (keyboard-quit))
#+END_SRC
**** Keybinding
#+BEGIN_SRC emacs-lisp
(define-key evil-normal-state-map   (kbd "C-g") #'evil-keyboard-quit) 
(define-key evil-motion-state-map   (kbd "C-g") #'evil-keyboard-quit) 
(define-key evil-insert-state-map   (kbd "C-g") #'evil-keyboard-quit) 
(define-key evil-window-map         (kbd "C-g") #'evil-keyboard-quit) 
(define-key evil-operator-state-map (kbd "C-g") #'evil-keyboard-quit) 
#+END_SRC
* Org
** ox-hugo
#+BEGIN_SRC emacs-lisp
(with-eval-after-load 'ox
  (require 'ox-hugo))
#+END_SRC
** org-indent
#+BEGIN_SRC emacs-lisp
(setq org-startup-indented t)
#+END_SRC
** Common Lisp & Scheme
#+BEGIN_SRC emacs-lisp
     (org-babel-do-load-languages
  'org-babel-load-languages
  '((scheme . t)
   (emacs-lisp . t)
   (ruby . t)
   (R . t)
   (python . t)
   (C . t)
   (sh . t)))
#+END_SRC
* Editor
** Company
#+BEGIN_SRC emacs-lisp
(global-company-mode t)
#+END_SRC
** yasnippet
#+BEGIN_SRC emacs-lisp
(setq yas-snippet-dirs
'("~/dotfiles/snippets"
    ))
(yas-global-mode 1)
#+END_SRC
** Hungry-delete
#+BEGIN_SRC emacs-lisp
(require 'hungry-delete)
(global-hungry-delete-mode)
#+END_SRC
** Show-paren
#+BEGIN_SRC emacs-lisp
(add-hook 'emacs-lisp-mode-hook 'show-paren-mode)
#+END_SRC
* Better Defaults
** Auto Backup & Auto Save
#+BEGIN_SRC emacs-lisp
(setq make-backup-files nil)
(setq auto-save-default nil)
#+END_SRC
** Auto Revert
#+BEGIN_SRC emacs-lisp
(global-auto-revert-mode t)
#+END_SRC
** Popwin
#+BEGIN_SRC emacs-lisp
(require 'popwin)
(popwin-mode t)
#+END_SRC
** Dired-x
#+BEGIN_SRC emacs-lisp
(require 'dired-x)
(setq dired-dwin-target 1)
#+END_SRC
* Abbrev
#+BEGIN_SRC emacs-lisp
(abbrev-mode t)
(define-abbrev-table 'global-abbrev-table '(
					    ("8em" "emacs")
					    ))
#+END_SRC
* Language Support
** Common Lisp
#+BEGIN_SRC emacs-lisp
(setq inferior-lisp-program "/usr/bin/sbcl")
    (add-to-list 'load-path "/usr/local/bin/slime/")
    (require 'slime)
    (slime-setup)
(slime-setup '(slime-fancy))
#+END_SRC
** Scheme
#+BEGIN_SRC emacs-lisp
(setq scheme-program-name "scm")
#+END_SRC



.#config.org
diphia@VM-0-7-ubuntu.4655:1589620884

config.el
(package-initialize)
(setq package-check-signature nil)
(when (>= emacs-major-version 24)
     (require 'package)
     (package-initialize)
     (setq package-archives '(("gnu"   . "http://elpa.emacs-china.org/gnu/")
                      ("melpa" . "http://elpa.emacs-china.org/melpa/"))))

 ;; cl - Common Lisp Extension
 (require 'cl)

 ;; Add Packages
 (defvar my/packages '(
                company
                evil
                nlinum-relative
                yasnippet
                ox-hugo
                helm
                monokai-theme
                hungry-delete
                swiper
                counsel
                popwin
                evil-collection
                slime
                paredit
                geiser
                ) "Default packages")

 (setq package-selected-packages my/packages)

 (defun my/packages-installed-p ()
     (loop for pkg in my/packages
           when (not (package-installed-p pkg)) do (return nil)
           finally (return t)))

 (unless (my/packages-installed-p)
     (message "%s" "Refreshing package database...")
     (package-refresh-contents)
     (dolist (pkg my/packages)
       (when (not (package-installed-p pkg))
         (package-install pkg))))

(tool-bar-mode -1)
(menu-bar-mode -1)
(scroll-bar-mode -1)

(setq inhibit-splash-screen 1)

(require 'nlinum-relative)
(nlinum-relative-setup-evil)                    ;; setup for evil
(add-hook 'prog-mode-hook 'nlinum-relative-mode)
(add-hook 'text-mode-hook 'nlinum-relative-mode)
(setq nlinum-relative-current-symbol "")      ;; or "" for display current line number
(setq nlinum-relative-redisplay-delay 0)      ;; delay
(setq nlinum-relative-offset 0)                 ;; 1 if you want 0, 2, 3...

(setq evil-want-keybinding nil)
(setq evil-want-C-i-jump nil) ;; use tab in normal mode while editing org file
(evil-mode 1)

(evil-collection-init)

(ivy-mode 1)

(require 'recentf)
(recentf-mode 1)
(setq recentf-max-menu-items 25)
(global-set-key "\C-x\ \C-r" 'recentf-open-files)

(defun open-init-file()
  (interactive)
  (find-file "~/dotfiles/config.org"))
(global-set-key (kbd "<f2>") 'open-init-file)

(global-set-key (kbd "C-w") 'backward-kill-word)

(global-set-key (kbd "C-h") 'delete-backward-char)

(global-set-key "\C-s" 'swiper)

(global-set-key (kbd "M-x") 'counsel-M-x)

(global-set-key (kbd "C-c p f") 'counsel-git)

(defun evil-keyboard-quit ()
  "Keyboard quit and force normal state."
  (interactive)
  (and evil-mode (evil-force-normal-state))
  (keyboard-quit))

(define-key evil-normal-state-map   (kbd "C-g") #'evil-keyboard-quit) 
(define-key evil-motion-state-map   (kbd "C-g") #'evil-keyboard-quit) 
(define-key evil-insert-state-map   (kbd "C-g") #'evil-keyboard-quit) 
(define-key evil-window-map         (kbd "C-g") #'evil-keyboard-quit) 
(define-key evil-operator-state-map (kbd "C-g") #'evil-keyboard-quit)

(with-eval-after-load 'ox
  (require 'ox-hugo))

(setq org-startup-indented t)

(org-babel-do-load-languages
'org-babel-load-languages
'((scheme . t)
 (emacs-lisp . t)
 (ruby . t)
 (R . t)
 (python . t)
 (C . t)
 (sh . t)))

(global-company-mode t)

(setq yas-snippet-dirs
'("~/dotfiles/snippets"
    ))
(yas-global-mode 1)

(require 'hungry-delete)
(global-hungry-delete-mode)

(add-hook 'emacs-lisp-mode-hook 'show-paren-mode)

(setq make-backup-files nil)
(setq auto-save-default nil)

(global-auto-revert-mode t)

(require 'popwin)
(popwin-mode t)

(require 'dired-x)
(setq dired-dwin-target 1)

(abbrev-mode t)
(define-abbrev-table 'global-abbrev-table '(
					    ("8em" "emacs")
					    ))

(setq inferior-lisp-program "/usr/bin/sbcl")
    (add-to-list 'load-path "/usr/local/bin/slime/")
    (require 'slime)
    (slime-setup)
(slime-setup '(slime-fancy))

(setq scheme-program-name "scm")
